src/System.Linq.Expressions/src/System/Linq/Expressions/MethodCallExpression.cs

```diff
@@ -879,6 +1042,34 @@ public static MethodCallExpression Call(Type type, string methodName, Type[] typ
         ///<paramref name="instance" />.Type is not assignable to the declaring type of the method represented by <paramref name="method" />.-or-The number of elements in <paramref name="arguments" /> does not equal the number of parameters for the method represented by <paramref name="method" />.-or-One or more of the elements of <paramref name="arguments" /> is not assignable to the corresponding parameter for the method represented by <paramref name="method" />.</exception>
         public static MethodCallExpression Call(Expression instance, MethodInfo method, IEnumerable<Expression> arguments)
         {
+            IReadOnlyList<Expression> argumentList = arguments as IReadOnlyList<Expression>;
```

Comment:
Maybe someone from the JIT team could chime in here? I can run a small test too.

Sent from my Windows Phone
________________________________
From: Stephen Toub<mailto:notifications@github.com>
Sent: ‎9/‎17/‎2015 3:58 PM
To: dotnet/corefx<mailto:corefx@noreply.github.com>
Cc: Bart De Smet<mailto:bartde@microsoft.com>
Subject: Re: [corefx] Use optimized N-ary MethodCallExpression nodes from all Call factories (#3288)


In src/System.Linq.Expressions/src/System/Linq/Expressions/MethodCallExpression.cs<https://na01.safelinks.protection.outlook.com/?url=https%3a%2f%2fgithub.com%2fdotnet%2fcorefx%2fpull%2f3288%23discussion_r39809036&data=01%7c01%7cbartde%40microsoft.com%7c82740a3ae3eb46c091dc08d2bfb37d5f%7c72f988bf86f141af91ab2d7cd011db47%7c1&sdata=KLYxZsPcWRNKw%2fChFFjMN3k05ensHUZqoyEH5%2bCziIM%3d>:

> @@ -879,6 +1042,34 @@ public static MethodCallExpression Call(Type type, string methodName, Type[] typ
>          ///<paramref name="instance" />.Type is not assignable to the declaring type of the method represented by <paramref name="method" />.-or-The number of elements in <paramref name="arguments" /> does not equal the number of parameters for the method represented by <paramref name="method" />.-or-One or more of the elements of <paramref name="arguments" /> is not assignable to the corresponding parameter for the method represented by <paramref name="method" />.</exception>
>          public static MethodCallExpression Call(Expression instance, MethodInfo method, IEnumerable<Expression> arguments)
>          {
> +            IReadOnlyList<Expression> argumentList = arguments as IReadOnlyList<Expression>;


I forget the details, but isn't there some kind of perf impact with regards to doing casts to interfaces with covariant type arguments?

—
Reply to this email directly or view it on GitHub<https://na01.safelinks.protection.outlook.com/?url=https%3a%2f%2fgithub.com%2fdotnet%2fcorefx%2fpull%2f3288%2ffiles%23r39809036&data=01%7c01%7cbartde%40microsoft.com%7c82740a3ae3eb46c091dc08d2bfb37d5f%7c72f988bf86f141af91ab2d7cd011db47%7c1&sdata=D%2fNMgZgK657ec15Dd56NXmEVl4j%2bRbqH9qJFd%2f%2fhkvY%3d>.